### ISSUE TITLE
SSL-related bug with websocket send call and multithreading

### ISSUE THREAD

**Issue by vpaeder** [Lawouach/WebSocket-for-Python#174]
hello everyone,
I bumped into a problem with the `send` method of `ws4py.websocket.WebSocket`when SSL is enabled.
I have a CherryPy server to which I associate some WebSocket servers. Messages are sent from a few independent threads through the same socket. This works fine without SSL.
However, with SSL enabled and more than one thread per socket, some messages end up being packed together and this makes the socket terminate.
If I open separate sockets for each thread, this works fine though.
I suspect there's something similar in essence as with [this bug](https://github.com/Lawouach/WebSocket-for-Python/issues/167), i.e. difference on how SSL and non-SSL sockets treat pending data (correct me if I didn't get it right).
Any clues on why this may happen and how to fix it?

**Comment by EternityForest** [comment]
Yeah, I'm on linux and didn't have much trouble testing this. It shows up only on SSL for me and goes away if I wrap part of send in a lock. I have a branch on my fork here:

https://github.com/EternityForest/WebSocket-for-Python/commit/f90b260710554d621f343300c7ac0cf7237be2bb

But I don't know quite enough about ws4py to be sure there's not other problems here.

**Comment by vpaeder** [comment]
Ok yes, this makes sense. If was trying to fiddle with the SSL_WANT_WRITE flag but didn't quite figure out how to integrate it cleanly at that point.
Thanks for the patch.
Works here too.

**Comment by dalai4git** [comment]
Not sure if I was experiencing the same problem, but locking didn't help in my case. For me the problem seemed to be that after the SSL_WANT_WRITE error, no retry was done. This seems to be a bug in `sendall` used in `ws4py.websocket.WebSocket._write`. See also https://github.com/pyca/pyopenssl/issues/176.

I just replaced `sendall` with `send` like in [this example](http://stackoverflow.com/questions/20715408/solved-pyopenssl-ssl3-write-pendingbad-write-retry-return-self-sslobj-write). This seems to work for me.

**Comment by dalai4git** [comment]
The solution above still has a problem when sending large messages (>16KB). Replacing the _write method in websocket with the following seems to work.

    def _write(self, b):
        if self.terminated or self.sock is None:
            raise RuntimeError("Cannot send on a terminated websocket")

        left_to_send = len(b)
        total_sent = 0
        while True:
            try:
                total_sent += self.sock.send(b)
                break
            except WantWriteError:
                pass

        left_to_send -= total_sent
        if left_to_send:
            self._write(b[total_sent:])
